### PR TITLE
Topic/leak join

### LINKED ISSUE #834: concurrent.join leaks memory
concurrent.join seems to leak memory, as following program will consume memory that retains in oldGen after manual GC. 

```scala 

object StreamsTestApp extends App {
  import Resources._

  def innerStream(idx:Int):Stream[Task, Unit] = {
    val allocate = "*"*10000
    Stream.eval(Task.delay(println(s"XXXG START $idx" + allocate.take(1))))
  }

  concurrent.join(Int.MaxValue)(
    time.sleep_[Task](10.seconds) ++
    Stream.range(0, 7000).map { i =>
      innerStream(i)
    } ++ Stream.eval_(Task.delay(println("XXXY DONE"))) ++
      time.sleep_[Task](1.hour)
  ).run.unsafeAttemptRun()


//  (
//    time.sleep_[Task](10.seconds) ++
//    Stream.range(0, 7000).flatMap { i =>
//      innerStream(i)
//    } ++ Stream.eval_(Task.delay(println("XXXY DONE"))) ++
//      time.sleep_[Task](1.hour)
//  ).run.unsafeAttemptRun()

}


```

After analysing through YourKit, the JVM seems to hold exactly 7000 references to stream in heap, together with their definition. 

The commented (sequential) variant is just working fine w/o leaking the memory. 

Any ideas of the source ?

### PR DESCRIPTION
**Pull request by pchlupacek**
@mpilquist @pchiusano @djspiewak this fixes #834. I'd like to make 0.9.5 with this fix today, if possible. Thanks

### CODE REVIEW

File: core/jvm/src/test/scala/fs2/ConcurrentSpec.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ class ConcurrentSpec extends Fs2Spec {
     }
 
     "join (1)" in forAll { (s1: PureStream[Int]) =>
-      runLog { concurrent.join(1)(s1.get.covary[Task].map(Stream.emit)) } shouldBe runLog { s1.get }
+      runLog { concurrent.join(1)(s1.get.covary[Task].map(Stream.emit)) }.toSet shouldBe runLog { s1.get }.toSet
```
Comment (mpilquist) [inline core/jvm/src/test/scala/fs2/ConcurrentSpec.scala]: Why this change?

**Comment by mpilquist** [comment]
Looks good to me though this is very much like the original version. There were subtle timing bugs in the original version. The current implementation tracked the inner streams running via resources (`Pull.acquireCancellable`) to prevent those leaks -- this was probably overkill. 

Given the importance of `concurrent.join`, I'm a little uncomfortable releasing 0.9.5 with a rewritten version without putting some more mileage on it. @pchlupacek -- can you run the join tests with 10x or 100x number of test cases?

**Comment by pchlupacek** [comment]
@mpilquist yes. However if you find a way how to crack current implementation, feel free. I had trouble to find a way to do it, but perhaps as you suggest is maybe one-liner.

File: core/jvm/src/test/scala/fs2/ConcurrentSpec.scala (hunk 1 above)
Comment (pchlupacek) [inline core/jvm/src/test/scala/fs2/ConcurrentSpec.scala]: sorry this is just overlook from my side. I did some experiments while implementing. Because maxopen is 1, the toSet is not necessary.

**Comment by pchlupacek** [comment]
@mpilquist just runn full fs2.test scope 20 times, so far no issue. Will continue up to 100 times and lets see if that proves fine too.

**Comment by pchlupacek** [comment]
Ok I have run 100 iteration of the test and all them passed

**Comment by pchlupacek** [comment]
@mpilquist if you afraid about implication of this change how about publishing 0.9.4.1 for the while and have it under some stress in our production code? Then after week or so we may give it a clear go to 0.9.5 ?

**Comment by mpilquist** [comment]
Any opinions on which PR we merge?

**Comment by mpilquist** [comment]
@pchlupacek OK, let's merge this and then I'll release 0.9.5-RC1. If that looks good after a few days in your production stuff, we can release 0.9.5.
